### PR TITLE
Pwm and individual contactor

### PR DESCRIPTION
**Pull request by connorl309**
# Quality Assurance Checklist 
To make reviews more efficient, please make sure the software feature meets the following standards and check everything off that meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions. 

## Requirements 
- [ ] Followed Coding Style Guide 
- [ ] Code Build checks pass 
- [ ] No merge conflicts 
- [x] Software feature has test associated with it. 
- [x] Software feature has documentation for it updated.
- [x] Test provides useful information and uses relevant data to accurately represent BPS 
    - NOTE: If test file already exists, use that one 
- [ ] 2 Members of the team have already reviewed my PR 
    - NOTE: Approver will not look at PR if 2 other members haven't reviewed first 
- [x] Did you Have Fun?

## Things to Consider 
- [ ] Even if the above are checked, is this the best way of writing my code? 
    - It's possible to write code that works, but are there ways to make code more efficient?

### CODE REVIEW

**Review by Champers5000 — APPROVED** [review]
Looks good to me, merge conflicts need to be fixed tho

**Comment by manthanand** [comment]
You will need to fix merge conflicts.

**Comment by connorl309** [comment]
> You will need to fix merge conflicts.

Yep working on it now

**Review by Champers5000 — DISMISSED** [review]
I don't see no problem no more ¯\(ㆆᴗㆆ)/¯

File: Apps/Inc/Contactor.h [hunk 1]
```diff
@@ -3,12 +3,13 @@
...
-#include "Simulator.h"
-#include "common.h"
+#include "BSP_PWM.h" //PWM_Period is defined in here
 
 //Parameters for PWM
-#define PWM_ON_TIME 2000 //This number is determined through testing
+#define PWM_ON_TIME 3500
+
+#include "Simulator.h"
+#include "common.h"
```
Comment (manthanand) [inline Apps/Inc/Contactor.h:12]: These lines didnt have to move. Only thing that should have changed is the PWM_ON_TIME

File: Apps/Inc/Fans.h [hunk 2]
```diff
@@ -22,17 +21,17 @@ Controls speed of fans which cool down the battery pack. Uses Pins PC6, 7 and PB
...
 /**
- * @brief This function will change the speed of the fans
- * @param fan Number of fan to change speed (1-4)
- * @param speed Speed of Fan(0-TOPSPEED)
- * @return ErrorStatus will return 1 if successful, 0 if error occurred
+ * @brief   Sets fan duty cycle
+ * @param   dutyCycle: int for duty cycle amount in range 0-8
+ * @param   fan: fan number whose speed should be changed
+ * @return  ErrorStatus
  */
```
Comment (manthanand) [inline Apps/Inc/Fans.h:28]: These comments are wrong and didnt have to be changed

File: Apps/Inc/Fans.h [hunk 3]
```diff
@@ -22,17 +21,17 @@ Controls speed of fans which cool down the battery pack. Uses Pins PC6, 7 and PB
...
 ErrorStatus Fans_Set(uint8_t fan, uint32_t speed);
 
 /**
- * @brief   Reads the duty cycle for a specified fan
- * @param   fan Gets the duty cycle for this fan
- * @return  int representation of the fan speed from 0-8, -1 if an error occurred
+ * @brief   Get current speed of specific fan
+ * @param   fan Number
+ * @return  Current PWM duty cycle if fans
  */
```
Comment (manthanand) [inline Apps/Inc/Fans.h:35]: Same as commment above

File: Apps/Src/Contactor.c [hunk 4]
```diff
@@ -8,11 +8,8 @@
  * @return none
  */
 void Contactor_Init(void) {
-    //Contactor Init is called before fan init, so initialize pwm here first. This will setup the contactor outputs
-    #ifdef SIMULATION
-        Simulator_Log(LOG_INFO, "Contactor Initialized\n");
-    #endif
-    BSP_PWM_Init();
+	//Contactor Init is called before fan init, so initialize pwm here first. This will setup the contactor outputs
+	BSP_PWM_Init();
```
Comment (manthanand) [inline Apps/Src/Contactor.c:12]: Simulator stuff shouldnt be removed.

File: Apps/Src/Contactor.c [hunk 5]
```diff
@@ -22,25 +19,14 @@ void Contactor_Init(void) {
...
-        break;
-    }
+	// set output pins to start outputing with a duty cycle set by PWM_ON_TIME in the header file
+	if(contactorChoice == ALL_CONTACTORS){
+		//c1 and cfan are the only two possible contactors, so we operate on both
+		BSP_PWM_Set(CFAN, PWM_ON_TIME);
+		BSP_PWM_Set(C1, PWM_ON_TIME);
+	}else{
+		BSP_PWM_Set(contactorChoice, PWM_ON_TIME);
+	}
```
Comment (manthanand) [inline Apps/Src/Contactor.c:29]: You can't just put in contactorChoice because that is a number from 1-4. I think this code would be better if you either had the enum for ``Contactors_t`` or the #defines for ``X_CONTACTOR_OUT`` in order to prevent this confusion.

File: Apps/Src/Contactor.c [hunk 6]
```diff
@@ -50,52 +36,21 @@ void Contactor_On(Contactors_t contactorChoice) {
...
-        default:
-            break;
-        }
+	if(contactorChoice == ALL_CONTACTORS){
+		//c1 and cfan are the only two possible contactors, so we operate on both
+		BSP_PWM_Set(CFAN, 0);
+		BSP_PWM_Set(C1, 0);
+	}else{
+		BSP_PWM_Set(contactorChoice, 0);
+	}
```
Comment (manthanand) [inline Apps/Src/Contactor.c:45]: Same as above comment

File: Apps/Src/Contactor.c [hunk 7]
```diff
@@ -50,52 +36,21 @@ void Contactor_On(Contactors_t contactorChoice) {
...
-		default:
-			retval = BSP_PWM_Get(HVLOW_CONTACTOR_OUT) || 
-					BSP_PWM_Get(HVHIGH_CONTACTOR_OUT) ||
-					BSP_PWM_Get(ARRAY_CONTACTOR_OUT);
-			break;
-		}
-        
-    return retval;
-}
+	return Contactor_Get((uint8_t) contactorChoice);
```
Comment (manthanand) [inline Apps/Src/Contactor.c:55]: This should not have been changed, we shouldn't be having Contactor code in the BSP_PWM library.

File: Apps/Src/Fans.c [hunk 8]
```diff
@@ -8,31 +8,26 @@ Uses Pins PC6,7 and PB14,15
...
- * @brief   Initialize all the GPIO pins connected to each Fan
- * @param   None
- * @return  None
- */
 void Fans_Init(void){
-    #ifdef SIMULATION
-        Simulator_Log(LOG_INFO, "Fans Initialized\n");
-    #endif
-    BSP_PWM_Init();
+    Fans_SetAll(4); //start with all fans half speed
```
Comment (manthanand) [inline Apps/Src/Fans.c:12]: ``Fans_Init()`` should not be setting the fans. I don't think anything here should have changed at all.

File: Apps/Src/Fans.c [hunk 9]
```diff
@@ -8,31 +8,26 @@ Uses Pins PC6,7 and PB14,15
...
 }
 
 /**
  * @brief This function will change the speed of the fans
  * @param fan Number of fan to change speed (1-4)
- * @param speed Speed of Fan(0-TOPSPEED)
+ * @param speed Speed of Fan(0-8)
```
Comment (manthanand) [inline Apps/Src/Fans.c:18]: This shouldn't have changed

File: Apps/Src/Fans.c [hunk 10]
```diff
@@ -8,31 +8,26 @@ Uses Pins PC6,7 and PB14,15
...
  * @return ErrorStatus will return 1 if successful, 0 if error occurred
  */
+
 ErrorStatus Fans_Set(uint8_t fan, uint32_t speed){
+    //don't mess with the contactor
+    BSP_PWM_Init();
```
Comment (manthanand) [inline Apps/Src/Fans.c:24]: Why is this called?

File: Apps/Src/Fans.c [hunk 11]
```diff
@@ -8,31 +8,26 @@ Uses Pins PC6,7 and PB14,15
...
     //don't mess with the contactor
     if (fan == CFAN) return ERROR;
     //Range of pulse is 0-4000
     //First check to make sure that change is within range of values
     //Load new value into Compare and Capture Register
-    return (speed < 0 || speed > TOPSPEED) ? BSP_PWM_Set(fan, TOPSPEED * DIVIDER) : BSP_PWM_Set(fan, speed * DIVIDER); //if not in range, leave function with error code
+    return (speed < 0 || speed > TOPSPEED) ? ERROR : BSP_PWM_Set(fan, speed * DIVIDER); //if not in range, leave function with error code
```
Comment (manthanand) [inline Apps/Src/Fans.c:30]: This shouldn't be changed

File: Apps/Src/Fans.c [hunk 12]
```diff
@@ -52,11 +47,10 @@ int Fans_GetSpeed(uint8_t fan){
  * @return  ErrorStatus
  */
 ErrorStatus Fans_SetAll(uint32_t speed) {
-    ErrorStatus result = SUCCESS;
+    ErrorStatus result;
```
Comment (manthanand) [inline Apps/Src/Fans.c:50]: This shouldn't have changed

File: BSP/Inc/BSP_PWM.h [hunk 13]
```diff
@@ -32,4 +36,12 @@ ErrorStatus BSP_PWM_Set(uint8_t pin, uint32_t speed);
...
 int BSP_PWM_Get(uint8_t pin);
 
+#define C1_PORT GPIOB //TO MAKE THE CONTACTOR CODE WORK, REMOVE ONCE IT HAS A NEW HOME. This is the port for the input pin
+/**
+ * @brief   Gets the state of the Contactor switch from one of its AUX pins.
+ * @note	THIS IS ALSO CODE THAT HAS NO HOME. You cannot get the state of ALL_CONTACTORS. As such, if that param is passed, it will return the state of the array contactor.
+ * @param   Contactor to get state of
+ * @return  0 if contactor is off/open, 1 if on/closed
+ */
+bool Contactor_Get(uint8_t contactorChoice);
```
Comment (manthanand) [inline BSP/Inc/BSP_PWM.h:46]: This shouldn't be here

File: BSP/STM32F413/Src/BSP_PWM.c [hunk 14]
```diff
@@ -85,17 +86,19 @@ void BSP_PWM_Init(void){
...
     GPIO_InitTypeDef GPIO_C1Init;
     GPIO_C1Init.GPIO_Pin = GPIO_Pin_1; //input pin is gpio B1
+
     GPIO_C1Init.GPIO_Mode = GPIO_Mode_IN;
     GPIO_C1Init.GPIO_Speed = GPIO_Speed_2MHz;
     GPIO_C1Init.GPIO_PuPd = GPIO_PuPd_DOWN;
     GPIO_C1Init.GPIO_OType = GPIO_OType_PP;
-    GPIO_Init(GPIOB, &GPIO_C1Init);
+
+    GPIO_Init(C1_PORT, &GPIO_C1Init);
```
Comment (manthanand) [inline BSP/STM32F413/Src/BSP_PWM.c:101]: this shouldn't be changed

File: BSP/STM32F413/Src/BSP_PWM.c [hunk 15]
```diff
@@ -166,3 +169,32 @@ int BSP_PWM_Get(uint8_t pin){
...
+		contactorReturnValue = ((C3_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1;
+	}
+	else if (contactorChoice == ALL_CONTACTORS) {
+		// return if ANY of the contactors are off.
+		contactorReturnValue = (((C1_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1)
+			&& (((C2_PORT->IDR & GPIO_Pin_5) >> 5) ? 0 : 1)
+			&& (((C3_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1);
+	}
+	*/
+	return contactorReturnValue;
```
Comment (manthanand) [inline BSP/STM32F413/Src/BSP_PWM.c:199]: This should be deleted

File: BSP/Simulator/Src/BSP_PWM.c [hunk 16]
```diff
@@ -81,3 +82,13 @@ int BSP_PWM_Get(uint8_t pin){
...
+
+/**
+ * @brief   Gets the state of the Contactor switch from one of its AUX pins.
+ * @note	THIS IS ALSO CODE THAT HAS NO HOME. You cannot get the state of ALL_CONTACTORS. As such, if that param is passed, it will return the state of the array contactor.
+ * @param   Contactor to get state of
+ * @return  0 if contactor is off/open, 1 if on/closed
+ */
+bool Contactor_Get(uint8_t contactorChoice) {
+    return pinSpeeds[4] != 0;
+}
```
Comment (manthanand) [inline BSP/Simulator/Src/BSP_PWM.c:94]: This should be deleted

File: Makefile [hunk 17]
```diff
@@ -29,22 +29,22 @@ flash:
...
+	@echo -e "	Set TEST only if you want to build a test."
+	@echo -e "	Otherwise, don't include TEST in the command."
+	@echo -e "	To build a test, replace ${PURPLE}<Test type>${NC} with the name of the file"
+	@echo -e "	excluding the file type (.c) e.g. say you want to test Voltage.c, call"
+	@echo -e -e "		${ORANGE}make ${BLUE}stm32f413 ${ORANGE}TEST=${PURPLE}Voltage${NC}"
+	@echo -e ""
+	@echo -e "When building for the board, you can specify custom define values for config.h."
+	@echo -e "See Config/Inc/config.h for all values."
+	@echo -e "Specify with the following ${PURPLE}format${NC}(quotes ARE needed):"
+	@echo -e "		${ORANGE}make ${BLUE}DEFINES=${GREEN}\"-D VARIABLE=VALUE VARIABLE=VALUE\"\n"
```
Comment (manthanand) [inline Makefile:47]: Why was this changed in this PR?

File: BSP/STM32F413/Makefile [hunk 18]
```diff
@@ -239,13 +239,13 @@ $(BUILD_DIR):
...
-	@echo "${GREEN}Compiling C files...${NC}"
+	@echo -e "${GREEN}Compiling C files...${NC}"
 
 asm_compile_msg:
-	@echo "${GREEN}Compiling ASM files...${NC}"
+	@echo -e "${GREEN}Compiling ASM files...${NC}"
 
 build_msg:
-	@echo "${CYAN}Generating BPS executable...${NC}"
+	@echo -e "${CYAN}Generating BPS executable...${NC}"
```
Comment (manthanand) [inline BSP/STM32F413/Makefile:248]: Why was this changed in this PR?

File: Tasks/Src/Task_FaultState.c [hunk 19]
```diff
@@ -123,6 +123,7 @@ void EnterFaultState() {
             sprintf(err, "$$$ Entered fault in state {%d} - ELECTRICAL STOP\n", stateCount - 1);
             Simulator_Log_Location(LOG_INFO, err);
         #endif
+            Contactor_Off(ALL_CONTACTORS);
```
Comment (manthanand) [inline Tasks/Src/Task_FaultState.c:126]: Why is this called again?

File: Validation/parse_simulator.py [hunk 20]
```diff
@@ -13,8 +13,7 @@
 PERIPHERALS: List[str] = [
     "ADC",
     "CAN",
-    "Contactor",
-    "Fan",
+    "PWM",
```
Comment (manthanand) [inline Validation/parse_simulator.py:16]: Why is this changed?

**Review by manthanand — CHANGES_REQUESTED** [review]

File: BSP/STM32F413/Makefile (hunk 18 above)
Comment (Champers5000) [reply BSP/STM32F413/Makefile:248]: make the makefile colors actually work right

File: Validation/parse_simulator.py (hunk 20 above)
Comment (Champers5000) [reply Validation/parse_simulator.py:16]: this is because the contactor and fan are no longer printing anything in the simulator when they are initialized, so this change was needed for the simulator check to pass

File: BSP/STM32F413/Src/BSP_PWM.c [hunk 21]
```diff
@@ -21,9 +21,9 @@ void BSP_PWM_Init(void){
     RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM12, ENABLE);
 
     //Enable Port Clocks
-    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOA, ENABLE);
-    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOB, ENABLE); //HV Contactor
-    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC, ENABLE);
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOA, ENABLE);
```
Comment (manthanand) [inline BSP/STM32F413/Src/BSP_PWM.c:24]: Could you undo all these changes.

File: Apps/Inc/Contactor.h [hunk 22]
```diff
@@ -17,11 +19,7 @@
 
 //These are defining how the contactors are wired, the array contactor is assumed to be wired to the fan board
 #define ARRAY_CONTACTOR_OUT CFAN
-#define HVHIGH_CONTACTOR_OUT C1
-#define HVLOW_CONTACTOR_OUT C1
-
-typedef uint8_t CONT_CHOICE;
-typedef enum {ARRAY_CONTACTOR = 1, HVHIGH_CONTACTOR = 2, HVLOW_CONTACTOR = 4, ALL_CONTACTORS = 7} Contactors_t;
+#define HV_CONTACTOR_OUT C1
```
Comment (manthanand) [inline Apps/Inc/Contactor.h:22]: In the event that we have 3 contactors. I would like to still keep ``HVLOW_CONTACTOR_OUT`` and just have it equal to ``C1``

File: Apps/Src/Contactor.c [hunk 23]
```diff
@@ -21,53 +21,19 @@ void Contactor_Init(void) {
...
-        BSP_PWM_Set(HVLOW_CONTACTOR_OUT, PWM_ON_TIME);
-        break;
-    case ALL_CONTACTORS: //c1 and cfan are the only two possible contactors, so we operate on both
-        BSP_PWM_Set(CFAN, PWM_ON_TIME);
-        BSP_PWM_Set(C1, PWM_ON_TIME);
-        break;
-    default:
-        break;
-    }
+    BSP_PWM_Set(contactorChoice, PWM_ON_TIME);
```
Comment (manthanand) [inline Apps/Src/Contactor.c:26]: Have a check to make sure it is a valid input.

File: Apps/Src/Contactor.c [hunk 24]
```diff
@@ -21,53 +21,19 @@ void Contactor_Init(void) {
...
-            BSP_PWM_Set(HVLOW_CONTACTOR_OUT, 0);
-            break;
-        case ALL_CONTACTORS:
-            BSP_PWM_Set(CFAN, 0);
-            BSP_PWM_Set(C1, 0);
-            break;
-        default:
-            break;
-        }
+    BSP_PWM_Set(contactorChoice, 0);
```
Comment (manthanand) [inline Apps/Src/Contactor.c:36]: Same as above comment.

**Review by manthanand — APPROVED** [review]
